applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -861,6 +954,647 @@ int handle_at_carrier(enum at_cmd_type cmd_type)
 	return ret;
 }
 
+#if defined(CONFIG_LWM2M_CARRIER_SETTINGS)
+/**
+ * @brief API to handle AT#XCARRIERCFG command.
+ */
+int handle_at_carrier_cfg(enum at_cmd_type cmd_type)
+{
+	int ret;
+	char op_str[SLM_CARRIER_CFG_OP_STR_MAX];
+	int size = sizeof(op_str);
+
+	if (cmd_type != AT_CMD_TYPE_SET_COMMAND) {
+		return -EINVAL;
+	}
+
+	ret = util_string_get(&slm_at_param_list, 1, op_str, &size);
+	if (ret) {
+		return ret;
+	}
+
+	ret = -EINVAL;
+	for (int i = 0; i < CARRIER_CFG_OP_MAX; i++) {
+		if (slm_util_casecmp(op_str, cfg_op_list[i].op_str)) {
+			ret = cfg_op_list[i].handler();
+			break;
+		}
+	}
+
+	return ret;
+}
+
+/* AT#XCARRIERCFG="apn"[,<apn>] */
+static int do_cfg_apn(void)
+{
+	uint32_t count;
+
+	count = at_params_valid_count_get(&slm_at_param_list);
+	if (count == 2) {
+		rsp_send("\r\n#XCARRIERCFG: %s\r\n", lwm2m_settings_apn_get());
+		return 0;
+	}
+
+	char apn[64];
+	int ret, size = sizeof(apn);
+
+	ret = util_string_get(&slm_at_param_list, 2, apn, &size);
+	if (ret) {
+		return ret;
+	}
+
+	return lwm2m_settings_apn_set(apn);
+}
+
+/* AT#XCARRIERCFG="auto_startup"[,<0|1>] */
+static int do_cfg_auto_startup(void)
+{
+	uint32_t count;
+
+	count = at_params_valid_count_get(&slm_at_param_list);
+	if (count == 2) {
+		rsp_send("\r\n#XCARRIERCFG: %u\r\n", lwm2m_settings_auto_startup_get());
+		return 0;
+	}
+
+	int ret;
+	uint16_t auto_startup;
+
+	ret = at_params_unsigned_short_get(&slm_at_param_list, 2, &auto_startup);
+	if (ret) {
+		return ret;
+	}
+
+	if (auto_startup > 1) {
+		LOG_DBG("AT#XCARRIERCFG=\"auto_startup\" failed: must be 0 or 1");
+		return -EINVAL;
+	}
+
+	return lwm2m_settings_auto_startup_set(auto_startup);
+}
+
+/* AT#XCARRIERCFG="bootstrap_smartcard"[,<0|1>] */
+static int do_cfg_bootstrap_from_smartcard(void)
+{
+	uint32_t count;
+
+	count = at_params_valid_count_get(&slm_at_param_list);
+	if (count == 2) {
+		rsp_send("\r\n#XCARRIERCFG: %u\r\n", lwm2m_settings_bootstrap_from_smartcard_get());
+		return 0;
+	}
+
+	int ret;
+	uint16_t enabled;
+
+	ret = at_params_unsigned_short_get(&slm_at_param_list, 2, &enabled);
+	if (ret) {
+		return ret;
+	}
+
+	if (enabled > 1) {
+		LOG_DBG("AT#XCARRIERCFG=\"bootstrap_smartcard\" failed: must be 0 or 1");
+		return -EINVAL;
+	}
+
+	return lwm2m_settings_bootstrap_from_smartcard_set(enabled);
+}
+
+const static uint32_t carriers_enabled_map[] = {
+	LWM2M_CARRIER_GENERIC,
+	LWM2M_CARRIER_VERIZON,
+	LWM2M_CARRIER_ATT,
+	LWM2M_CARRIER_LG_UPLUS,
+	LWM2M_CARRIER_T_MOBILE,
+	LWM2M_CARRIER_SOFTBANK
+};
```

Comment:
With this will the CONFIG_LWM2M_CARRIER_**** be gone?

Reply:
Not necessarily gone, the Kconfigs will be overwritten only if the corresponding group of settings is enabled. For example, if `CARRIER_CFG_OP_DEVICE_CONFIG_ENABLE` is set to true, then all `CARRIER_CFG_OP_DEVICE_***` will apply. If set to false, whatever is set in Kconfig will apply instead. In case of `CARRIER_CFG_OP_CARRIERS`, this will be enabled or disabled through `CARRIER_CFG_OP_ENABLE`. I don't think I've made that one clear enough, so I'll rename it to show that it belongs to that group.